src/tools/remote-test-server/src/main.rs

```diff
@@ -202,6 +205,28 @@ fn handle_run(socket: TcpStream, work: &Path, lock: &Mutex<()>) {
     ]));
 }
 
+fn recv<B: BufRead>(dir: &Path, io: &mut B) -> PathBuf {
+    let mut filename = Vec::new();
+    t!(io.read_until(0, &mut filename));
+
+    // We've got some tests with *really* long names. We try to name the test
+    // executable the same on the target as it is on the host to aid with
+    // debugging, but the targets we're emulating are often more restrictive
+    // than the hosts as well.
+    //
+    // To ensure we can run a maximum number of tests without modifications we
+    // just arbitrarily truncate the filename to 50 bytes. That should
+    // hopefully allow us to still identify what's running while staying under
+    // the filesystem limits.
+    let len = cmp::min(filename.len() - 1, 50);
+    let dst = dir.join(t!(str::from_utf8(&filename[..len])));
```

Comment:
So a funny story about this logic as well. I was getting a random failure in a test with a super long test name and I couldn't figure out why. If I ran only that one test it passed, but in the entire suite it failed. Turned out:

* Running the test alone the path was 126 characters
* Running the test after 10 other tests the path was 127 characters
* Running the test after 100 other tests the pth was 128 characters

The test only ended up failing if it was run after 100 other tests. Sure enough each test claims itself a number and is run in a `testXX` directory, so after 100 tests we're in 3 character range, pushing us from 127 to 128.

Turns out Android has trouble with paths longer than 127 characters. Who knew!

All of this was because we *used* to run tests in `/data/tmp`, and I changed it in this PR to `/data/tmp/work`. Wheeeeeeeeeeeeeeee